From: Michal Cieslakiewicz <michal.cieslakiewicz@wp.pl>
Subject: [PATCH v5 5/8] mac80211: ath9k: enable GPIO buttons

Enable platform-defined GPIO button support for ath9k device.
Key poller is activated for attached platform buttons.
Requires ath9k GPIO chip access.

Signed-off-by: Michal Cieslakiewicz <michal.cieslakiewicz@wp.pl>
Signed-off-by: Felix Fietkau <nbd@nbd.name>
---
--- a/drivers/net/wireless/ath/ath9k/ath9k.h
+++ b/drivers/net/wireless/ath/ath9k/ath9k.h
@@ -1051,6 +1051,7 @@ struct ath_softc {
 #ifdef CONFIG_GPIOLIB
 	struct gpio_chip *gpiochip;
 	struct gpio_desc *gpiodesc;
+	struct platform_device *btnpdev;	/* gpio-keys-polled */
 #endif
 #endif
 
--- a/drivers/net/wireless/ath/ath9k/gpio.c
+++ b/drivers/net/wireless/ath/ath9k/gpio.c
@@ -15,6 +15,8 @@
  */
 
 #include "ath9k.h"
+#include <linux/platform_device.h>
+#include <linux/gpio_keys.h>
 
 #ifdef CPTCFG_MAC80211_LEDS
 
@@ -117,6 +119,67 @@ static void ath9k_unregister_gpio_chip(s
 	kfree(gc);
 }
 
+/******************/
+/*  GPIO Buttons  */
+/******************/
+
+/* add GPIO buttons */
+static void ath9k_init_buttons(struct ath_softc *sc)
+{
+	struct ath_hw *ah = sc->sc_ah;
+	struct platform_device *pdev;
+	struct gpio_keys_platform_data gkpdata;
+	struct gpio_keys_button *bt;
+	int i;
+
+	if (!sc->gpiochip)
+		return;
+
+	if (!ah->btns || !ah->num_btns)
+		return;
+
+	bt = devm_kmemdup(sc->dev, ah->btns,
+			  ah->num_btns * sizeof(struct gpio_keys_button),
+			  GFP_KERNEL);
+	if (!bt)
+		return;
+
+	for (i = 0; i < ah->num_btns; i++) {
+		if (ah->btns[i].gpio == sc->sc_ah->led_pin)
+				sc->sc_ah->led_pin = -1;
+
+		ath9k_hw_gpio_request_in(sc->sc_ah, ah->btns[i].gpio,
+					 "ath9k-gpio");
+		bt[i].gpio = sc->gpiochip->base + ah->btns[i].gpio;
+	}
+
+	memset(&gkpdata, 0, sizeof(struct gpio_keys_platform_data));
+	gkpdata.buttons = bt;
+	gkpdata.nbuttons = ah->num_btns;
+	gkpdata.poll_interval = ah->btn_poll_interval;
+
+	pdev = platform_device_register_data(sc->dev, "gpio-keys-polled",
+					     PLATFORM_DEVID_AUTO, &gkpdata,
+					     sizeof(gkpdata));
+	if (!IS_ERR_OR_NULL(pdev))
+		sc->btnpdev = pdev;
+	else {
+		sc->btnpdev = NULL;
+		devm_kfree(sc->dev, bt);
+	}
+}
+
+/* remove GPIO buttons */
+static void ath9k_deinit_buttons(struct ath_softc *sc)
+{
+	if (!sc->gpiochip || !sc->btnpdev)
+		return;
+
+	platform_device_unregister(sc->btnpdev);
+
+	sc->btnpdev = NULL;
+}
+
 #else /* CONFIG_GPIOLIB */
 
 static inline void ath9k_register_gpio_chip(struct ath_softc *sc)
@@ -127,6 +190,14 @@ static inline void ath9k_unregister_gpio
 {
 }
 
+static inline void ath9k_init_buttons(struct ath_softc *sc)
+{
+}
+
+static inline void ath9k_deinit_buttons(struct ath_softc *sc)
+{
+}
+
 #endif /* CONFIG_GPIOLIB */
 
 /********************************/
@@ -232,6 +303,7 @@ void ath_deinit_leds(struct ath_softc *s
 {
 	struct ath_led *led;
 
+	ath9k_deinit_buttons(sc);
 	while (!list_empty(&sc->leds)) {
 		led = list_first_entry(&sc->leds, struct ath_led, list);
 #ifdef CONFIG_GPIOLIB
@@ -267,6 +339,7 @@ void ath_init_leds(struct ath_softc *sc)
 	}
 
 	ath_fill_led_pin(sc);
+	ath9k_init_buttons(sc);
 
 	if (ah->led_pin < 0)
 		return;
